src/openms/source/CHEMISTRY/RibonucleotideDB.cpp

```diff
@@ -67,6 +73,141 @@ namespace OpenMS
+    {
+      ribo->setOrigin(string(entry.at("reference_moiety").at(0))[0]);
+      ribo->setTermSpecificity(Ribonucleotide::ANYWHERE); // due to format changes we get the terminal specificity from the moieties, modomics contains base specific terminals, but they can be represented by the wild-card ones
+    }
+    else if (entry["reference_moiety"].size() == 4) // if all moieties are possible it might be a terminal
+    {
+      ribo->setOrigin('X'); // Use X as any unmodified
+      if (code.hasSuffix("pN"))
+      {
+        ribo->setTermSpecificity(Ribonucleotide::FIVE_PRIME);
+      }
+      else if (code.hasSuffix("p") && code.hasPrefix("N"))
+      {
+        ribo->setTermSpecificity(Ribonucleotide::THREE_PRIME);
+      }
+      else
+      {
+        ribo->setTermSpecificity(Ribonucleotide::ANYWHERE); //other nonspecific mods
+      }
+    }
+    else
+    {
+      OPENMS_LOG_ERROR << "Error: we don't support bases with multiple reference moieties or multicharacter moieties." << endl;
+    }
+    ribo->setHTMLCode(entry.at("abbrev")); //This is the single letter unicode representation that only SOME mods have
+    ribo->setFormula(EmpiricalFormula(entry.at("formula")));
+    if (!(entry.at("mass_avg").is_null()))
+    {
+      ribo->setAvgMass(entry.at("mass_avg"));
+    }
+    if (ribo->getAvgMass() - ribo->getFormula().getAverageWeight() >= 0.01)
+    {
+      OPENMS_LOG_WARN << "Average mass of " << code << " differs substantially from its formula mass.\n";
+    }
+
+    if (!(entry.at("mass_monoiso").is_null()))
+    {
+      ribo->setMonoMass(entry.at("mass_monoiso"));
+    }
+    if (ribo->getMonoMass() - ribo->getFormula().getMonoWeight() >= 0.01)
+    {
+      OPENMS_LOG_WARN << "Average mass of " << code << " differs substantially from its formula mass.\n";
+    }
+
+    //TODO: Calculate base loss formula from SMILES
+    if (code.hasPrefix('d')) // handle deoxyribose, possibly with methyl mod
+    {
+      ribo->setBaselossFormula(EmpiricalFormula("C5H10O4"));
+    }
+    else if (code.hasSuffix('m')) // mod. attached to the ribose, not base
+    {
+      ribo->setBaselossFormula(EmpiricalFormula("C6H12O5"));
+    }
+    else if (code.hasSuffix("m*")) // check if we have both a sulfer and a 2'-O methyl
+    {
+      ribo->setBaselossFormula(EmpiricalFormula("C6H12O5"));
+    }
+    else if (code.hasSuffix('?') || code.hasSuffix("?*")) // ambiguity code -> fill the map
+    {
+      if (!entry.contains("alternatives"))
+      {
+        String msg = "Ambiguous mod without alternative found in " + code;
+        //throw Exception::ParseError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, code, msg);
+      }
+      String code1 = string(entry.at("alternatives").at(0)), code2 = string(entry.at("alternatives").at(1)); //we always have exactly two ambiguities
+      ambiguity_map_[code] = make_pair(getRibonucleotide(code1), getRibonucleotide(code2));
```

Comment:
@timosachsenberg any idea why I get a `not declared in this scope` error for `ambiguity_map_` its a protected member of RibonucleotideDB, and works fine in the `parseRow` function. I assume its something to do with how `parseEntry` was forward declared in this file and `parseRow` is declared in the header.

Reply:
parse entry is not a member